implement app state getter for web

navaronbracke opened the pull request:
Fixes https://github.com/auth0/auth0-flutter/issues/486
Replaces https://github.com/auth0/auth0-flutter/pull/485 (I had an issue with signing the commits during the interactive rebase, so I started over)

<!--
❗ For general support or usage questions, use the Auth0 Community forums or raise a support ticket.

By submitting a pull request to this repository, you agree to the terms within the Auth0 Code of Conduct: https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md.
-->

- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

<!--
❗ All the above items are required. Pull requests with an incomplete or missing checklist will be closed.
-->

### 📋 Changes

<!--
Describe both what is changing and why this is important. Include:

- Types and methods added, deleted, deprecated, or changed
- A summary of usage if this is a new feature or a change to a public API
-->

This PR adds support for the `appState` parameter to `loginWithRedirect()` for the web.
See https://github.com/auth0/auth0-spa-js/blob/f2e566849efa398ca599daf9ebdfbbd62fcb1894/src/global.ts#L298

There is prior art in https://github.com/auth0/auth0-angular/pull/168

### 🎯 Testing

<!--
Describe how this can be tested by reviewers. Be specific about anything not tested and why. Include any manual steps for testing end-to-end, or for testing functionality not covered by unit tests.
-->

Added several unit tests. This can be tested end-to-end by setting up a login with redirect, that uses the app state argument as parameter, and then validating that the app state is returned when the user is redirected back.
I did verify end-to-end using an internal application that uses the SDK integration.

To run the unit tests locally, use `flutter test test/web/auth0_flutter_web_test.dart --platform=chrome`

navaronbracke commented:
@Widcket Apologies, I had an issue during the interactive rebase when trying to fix the commit signing, so I had to start over. I also applied your comments from the previous PR into this one.

navaronbracke commented on auth0_flutter/lib/auth0_flutter_web.dart:
```diff
@@ -19,6 +19,12 @@ class Auth0Web {
+  /// Get the app state that was provided during a previous call
+  /// to [loginWithRedirect].
+  ///
+  /// This method should be called after calling [onLoad].
```
The note about second access returning null is now removed.

navaronbracke commented on auth0_flutter/lib/src/web/auth0_flutter_plugin_real.dart:
```diff
@@ -23,6 +23,21 @@ class Auth0FlutterPlugin extends Auth0FlutterWebPlatform {
+  /// Thus clearing this object is not needed,
+  /// as the actual state is managed across reloads,
+  /// using the transaction manager.
+  // TODO: move the `appState` to the result of `onLoad/initialize`
```
Please let me know if I need to update this new TODO

navaronbracke commented on auth0_flutter/lib/src/web/auth0_flutter_plugin_real.dart:
```diff
@@ -23,6 +23,21 @@ class Auth0FlutterPlugin extends Auth0FlutterWebPlatform {
+  Object? _appState;
+
+  @override
+  Future<Object?> get appState => Future<Object?>.value(_appState);
```
Per your request, the getter now returns the value as-is

navaronbracke commented on auth0_flutter_platform_interface/lib/src/auth0_flutter_web_platform.dart:
```diff
@@ -15,6 +15,10 @@ abstract class Auth0FlutterWebPlatform extends PlatformInterface {
     _instance = instance;
   }
 
+  Future<Object?> get appState {
```
Previously there was also an `@override` that duplicated this implementation. I removed that one.
Maybe that is why the code coverage report was reporting the uncovered lines?

navaronbracke commented on auth0_flutter/test/web/auth0_flutter_web_test.dart:
```diff
@@ -72,13 +74,91 @@ void main() {
+
+    when(mockClientProxy.isAuthenticated())
+        .thenAnswer((final _) => Future.value(false));
+    when(mockClientProxy.handleRedirectCallback())
```
This test was updated to verify the redirect callback as well.

navaronbracke commented on auth0_flutter/test/web/auth0_flutter_web_test.dart:
```diff
@@ -72,13 +74,91 @@ void main() {
+    );
+  });
+
+  test('appState getter returns value when accessed more than once', () async {
```
This test was added to prevent regressing the behavior of the new getter. (it should always return the same value)

navaronbracke commented:
It seems that the upstream iOS & MacOS tests are broken currently? This prevents new test reports from being made.

Widcket commented:
> It seems that the upstream iOS & MacOS tests are broken currently? This prevents new test reports from being made.

The smoke tests are, but the coverage is not taken from those but from the unit tests.

navaronbracke commented on auth0_flutter_platform_interface/lib/src/auth0_flutter_web_platform.dart:
```diff
@@ -1,7 +1,9 @@
 
-class StubAuth0FlutterWeb extends Auth0FlutterWebPlatform {}
+class StubAuth0FlutterWeb extends Auth0FlutterWebPlatform {
+  StubAuth0FlutterWeb() : super();
```
I added this line, so that the super invocation of `Auth0FlutterWebPlatform()` can be tested.

navaronbracke commented on auth0_flutter_platform_interface/test/auth0_flutter_web_platform_test.dart:
```diff
@@ -0,0 +1,90 @@
+import 'package:auth0_flutter_platform_interface/src/web/client_options.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+class TestAuth0FlutterWeb extends Auth0FlutterWebPlatform {
```
This file should fix the test coverage for `Auth0FlutterWebPlatform`.
I ran `flutter test --coverage` before adding these tests and the code coverage lcov.info didn't list anything for `Auth0FlutterWebPlatform`.

After adding these tests and running `flutter test --coverage` again, the lcov.info showed the class as fully tested.

Widcket replied:
Thanks for adding this.

navaronbracke commented:
@Widcket Code coverage should be fine now. PTAL.

navaronbracke commented:
@Widcket I had a look at the failing tests:

The failing iOS tests seem to be due to CI using XCode 15.0.1 ?

When I open the relevant code that is outlined in the logs

```
/Users/runner/work/auth0-flutter/auth0-flutter/auth0_flutter/example/ios/Pods/Auth0/Auth0/UIWindow+TopViewController.swift:14:40: error: cannot call value of non-function type 'UIApplication'
[1333](https://github.com/auth0/auth0-flutter/actions/runs/13494371876/job/37749723978?pr=509#step:4:1336)
        guard let root = UIApplication.shared()?.windows.last(where: \.isKeyWindow)?.rootViewController else {
[1334](https://github.com/auth0/auth0-flutter/actions/runs/13494371876/job/37749723978?pr=509#step:4:1337)
                                       ^     ~~
[1335](https://github.com/auth0/auth0-flutter/actions/runs/13494371876/job/37749723978?pr=509#step:4:1338)
                                             
[1336](https://github.com/auth0/auth0-flutter/actions/runs/13494371876/job/37749723978?pr=509#step:4:1339)
/Users/runner/work/auth0-flutter/auth0-flutter/auth0_flutter/example/ios/Pods/Auth0/Auth0/UIWindow+TopViewController.swift:14:70: error: cannot infer key path type from context; consider explicitly specifying a root type
[1337](https://github.com/auth0/auth0-flutter/actions/runs/13494371876/job/37749723978?pr=509#step:4:1340)
        guard let root = UIApplication.shared()?.windows.last(where: \.isKeyWindow)?.rootViewController else {
[1338](https://github.com/auth0/auth0-flutter/actions/runs/13494371876/job/37749723978?pr=509#step:4:1341)
                                                                     ^
```

I don't see any compilation warnings / errors when running this with XCode version 16.2 (16C5032a)

![Screenshot 2025-02-27 at 15 25 49](https://github.com/user-attachments/assets/3b016421-c631-4fc2-b3a8-78a066d226bd)


As for the failing macOS tests, there is a redundant `let` keyword, which can be removed to fix the warning:

![Screenshot 2025-02-27 at 15 33 19](https://github.com/user-attachments/assets/78200fc1-d9a9-4fa3-a022-542e9eea9a07)

Do you want me to commit the fix for the `let` keyword in this PR? Or will you update the XCode version for CI and fix the let keyword together with that?

Widcket commented:
>The failing iOS tests seem to be due to CI using XCode 15.0.1 ?
When I open the relevant code that is outlined in the logs

Thank you for taking a look. It should compile on Xcode 15.0.1, as the underlying Swift SDK (Auth0.swift) does. Also, we need to test using Xcode 15, as it's still possible to use it to compile apps to submit to the App Store. That is set to change on April:

<img width="981" alt="Screenshot 2025-02-27 at 14 56 53" src="https://github.com/user-attachments/assets/ad0072ed-73c0-4d2e-9ed2-3b32287126f2" />

navaronbracke commented:
Maybe we should update to the latest hotfix version for XCode 15, which is XCode 15.4? Maybe the fix for that problem is in one of the hotfixes for XCode 15 ?

Widcket commented:
That's a good idea, though I'll start right with 15.4, and if that fixes it, then try one version back at a time.

Widcket commented:
Nope, that didn't fix it https://github.com/auth0/auth0-flutter/actions/runs/13570466141/job/37938452729?pr=520

navaronbracke commented:
Well, now there is a different issue for iOS:

```
/Users/runner/work/auth0-flutter/auth0-flutter/auth0_flutter/example/ios/Pods/Auth0/Auth0/ASProvider.swift:70:47: error: cannot assign value of type 'ASUserAgent' to type '(any ASWebAuthenticationPresentationContextProviding)?'
        session.presentationContextProvider = self
                                              ^~~~
/Users/runner/work/auth0-flutter/auth0-flutter/auth0_flutter/example/ios/Pods/Auth0/Auth0/ASProvider.swift:70:47: note: add missing conformance to 'ASWebAuthenticationPresentationContextProviding' to class 'ASUserAgent'
        session.presentationContextProvider = self
                                              ^
```

Widcket commented:
> Well, now there is a different issue for iOS:

My bad. That's the issue I thought you meant could be fixed by updating Xcode. I see now that there is Ruby problem, and I just raised https://github.com/auth0/auth0-flutter/pull/523 to fix it.

Regarding `error: cannot assign value of type 'ASUserAgent' to type '(any ASWebAuthenticationPresentationContextProviding)?'`, it doesn't happen for Auth0.swift (it compiles fine) and only happens here when the GitHub actions get cached. I have to look deeper into it.

navaronbracke commented:
The native iOS unit tests seem to pass on your PR for the Ruby version? Interesting, maybe there is indeed a caching issue in Github Actions.

navaronbracke commented:
Since we now have WASM support, I am in the process of updating the web specific mocking for the tests.

navaronbracke commented:
@pmathew92 It turns out that the original guidance for mocking JS interop types is a bit outdated.
I managed to fix a first test case, by taking the existing interop tests from the `camera_web` package as an example (the first-party camera plugin). I'll be updating the other tests today and then I'll push that to this PR.

pmathew92 commented:
Hi @navaronbracke  , i have already fixed the tests when merging the WASM PR . Could you take the latest pull and try again or are you observing any new error?

navaronbracke commented:
I am referring to the web tests when running with JS interop extension types, not any of the native tests (iOS or MacOS).

pmathew92 commented:
Yes , i have fixed them too. Please check the latest master @navaronbracke

navaronbracke commented:
@pmathew92 I cannot seem to update the mocks locally?

```
navaronbracke@MacBook-Pro-van-Navaron auth0_flutter % flutter pub run build_runner build --delete-conflicting-outputs
Deprecated. Use `dart run` instead.
Building package executable... (4.4s)
Built build_runner:build_runner.
[INFO] Generating build script completed, took 426ms
[INFO] Precompiling build script... completed, took 4.1s
[INFO] Building new asset graph completed, took 1.4s
[INFO] Checking for unexpected pre-existing outputs. completed, took 2ms
[INFO] Generating SDK summary completed, took 6.6s
[SEVERE] mockito:mockBuilder on test/web/auth0_flutter_web_test.dart:

Bad state: Asset URI is missing for abstract class JSObject
package:mockito/src/builder.dart 2223:10                     _MockClassInfo._typeImport
package:mockito/src/builder.dart 1695:21                     _MockClassInfo._addFakeClass.<fn>.<fn>.<fn>
package:code_builder/src/specs/type_reference.g.dart 164:33  _$TypeReferenceBuilder.update
package:code_builder/src/specs/type_reference.g.dart 22:36   new _$TypeReference
package:mockito/src/builder.dart 1692:33                     _MockClassInfo._addFakeClass.<fn>.<fn>
package:mockito/src/builder.dart 2101:24                     _MockClassInfo._withTypeParameters
package:mockito/src/builder.dart 1689:7                      _MockClassInfo._addFakeClass.<fn>
package:code_builder/src/specs/class.g.dart 345:33           _$ClassBuilder.update
package:code_builder/src/specs/class.g.dart 40:28            new _$Class
package:mockito/src/builder.dart 1682:37                     _MockClassInfo._addFakeClass
package:mockito/src/builder.dart 1641:7                      _MockClassInfo._dummyFakedValue
package:mockito/src/builder.dart 1663:27                     _MockClassInfo._dummyValueImplementing
package:mockito/src/builder.dart 1568:12                     _MockClassInfo._dummyValue
package:mockito/src/builder.dart 1671:11                     _MockClassInfo._dummyValueImplementing
package:mockito/src/builder.dart 1568:12                     _MockClassInfo._dummyValue
package:mockito/src/builder.dart 1972:24                     _MockClassInfo._buildOverridingGetter
package:mockito/src/builder.dart 1236:36                     _MockClassInfo.fieldOverrides.<fn>
package:code_builder/src/specs/method.g.dart 323:33          _$MethodBuilder.update
package:code_builder/src/specs/method.g.dart 38:29           new _$Method
package:mockito/src/builder.dart 1236:15                     _MockClassInfo.fieldOverrides
dart:core                                                    List.addAll
package:built_collection/src/list/list_builder.dart 98:14    ListBuilder.addAll
package:mockito/src/builder.dart 1194:28                     _MockClassInfo._buildMockClass.<fn>.<fn>
package:mockito/src/builder.dart 2101:24                     _MockClassInfo._withTypeParameters
package:mockito/src/builder.dart 1148:7                      _MockClassInfo._buildMockClass.<fn>
package:code_builder/src/specs/class.g.dart 345:33           _$ClassBuilder.update
package:code_builder/src/specs/class.g.dart 40:28            new _$Class
package:mockito/src/builder.dart 1126:12                     _MockClassInfo._buildMockClass
package:mockito/src/builder.dart 1072:9                      new _MockLibraryInfo
package:mockito/src/builder.dart 94:29                       MockBuilder.build

[INFO] Running build completed, took 23.5s
[INFO] Caching finalized dependency graph completed, took 81ms
[SEVERE] Failed after 23.6s
Failed to update packages.
```

I am running on Dart 3.7.0 / Flutter 3.29.0 (regardless of the version in the pubspec, which is lower)

navaronbracke commented:
@pmathew92 Did you run the web tests with `flutter test test/web/auth0_flutter_web_test.dart --platform=chrome` when implementing WASM ? Because that doesn't work without updated mocks.

pmathew92 commented:
@navaronbracke  I ran `flutter test --tags browser --platform chrome` within the auth0_flutter module and it passed all the tests

navaronbracke commented:
I did manage to fix the mocks locally, seems to be an issue with build_runner / mockito. Had to do it manually though.
I still have a different issue now, but I'll get that sorted also. Once the tests pass on my end I will push an update here.

navaronbracke commented:
Tests passed locally now.
![Screenshot 2025-03-05 at 14 08 11](https://github.com/user-attachments/assets/bdf36025-0912-4c28-9fb9-bf99aadb1f60)

navaronbracke commented on auth0_flutter/test/web/extensions/web_exception_extension_test.dart:
```diff
@@ -22,11 +22,13 @@ void main() {
 
   test('additional props are added to the details collection', () {
     final exception = createJsException(
```
Formatting change (adding missing comma's). Some indents were off.

navaronbracke commented on auth0_flutter/test/web/extensions/credentials_extension_test.dart:
```diff
@@ -18,15 +18,18 @@ void main() {
       const expiresIn = 8400;
       const expectedTokenType = 'Bearer';
       final webCredentials = WebCredentials(
-          access_token: accessToken, id_token: idToken,
```
Formatting change (adding missing comma's). Some indents were off.

navaronbracke commented on auth0_flutter/darwin/Classes/AuthAPI/AuthAPIPhoneNumberPasswordlessLoginMethod.swift:
```diff
@@ -37,7 +37,7 @@ struct AuthAPIPasswordlessPhoneNumberMethodHandler: MethodHandler {
             .parameters(["authParams":parameters])
             .start {
                 switch $0 {
-                case let .success:
```
This was a warning that was surfaced in the MacOS CI job.

navaronbracke replied:
@pmathew92 The code coverage report lists this line as untested.
However, the entire `func handle(with arguments: [String: Any], callback: @escaping FlutterResult)` method is shown as untested. Are there plans to add tests for this method? Or can I ignore that for this PR?

pmathew92 replied:
@navaronbracke  You can ignore that for this PR

navaronbracke commented on auth0_flutter/lib/src/web/js_interop_utils.dart:
```diff
@@ -24,12 +24,30 @@ class JsInteropUtils {
+  /// that do not fit into a static interop definition.
+  ///
+  /// See https://api.dart.dev/dart-js_interop/JSAnyUtilityExtension/dartify.html
+  static Object? dartifyObject(final JSAny? obj) => obj.dartify();
```
I updated this to match the WASM compatible API (`JSAny?` type).

However @pmathew92 I see that some of the API used in this file needs Dart 3.6.0. Do we use a forwarding stub to fix that? Or do we update the pubspec?

![Screenshot 2025-03-05 at 14 14 45](https://github.com/user-attachments/assets/980021c8-5410-478e-850a-39a3e03072e3)

navaronbracke replied:
I fixed this with a polyfill-extension and added a regression test also.

pmathew92 replied:
@navaronbracke  Do we see any breaking changes if we updated to 3.6 ?

navaronbracke replied:
There are additions for JS interop (such as the ones I mentioned here)

https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#360

Flutter 3.27 also has some breaking changes:
https://docs.flutter.dev/release/breaking-changes#released-in-flutter-3-27

navaronbracke commented on auth0_flutter/test/web/auth0_extension_type_mocks.dart:
```diff
@@ -13,6 +13,11 @@ class Auth0ClientImpl {
 }
 
+@JSExport()
+class RedirectLoginResultImpl {
```
This is part of the fix for the updated mocks

navaronbracke commented on auth0_flutter/test/web/auth0_flutter_web_test.mocks.dart:
```diff
@@ -46,6 +48,18 @@ class _FakeWebCredentials_1 extends _i1.SmartFake
 }
 
+@JSExport()
+class _FakeRedirectLoginResult_2 extends _i1.SmartFake
```
I had to add this manually, since mockito would crash. Is a known issue, though.
https://github.com/dart-lang/mockito/issues/762

navaronbracke commented on auth0_flutter/test/web/auth0_flutter_web_test.mocks.dart:
```diff
@@ -107,20 +126,33 @@ class MockAuth0FlutterWebClientProxy extends _i1.Mock
-        returnValue: _i4.Future<void>.value(),
-        returnValueForMissingStub: _i4.Future<void>.value(),
-      ) as _i4.Future<void>);
+        returnValue: _i4.Future<_i2.RedirectLoginResult>.value(
```
I had to fix this mock manually also

navaronbracke commented on auth0_flutter/analysis_options.yaml:
```diff
@@ -32,7 +32,6 @@ linter:
     test_types_in_equals: true
     throw_in_finally: true
     unnecessary_statements: true
-    unsafe_html: true
```
`unsafe_html` is removed in Dart 3.7. The project had no ignores, so this can be removed.

navaronbracke commented on auth0_flutter/lib/src/web/js_interop_utils.dart:
```diff
@@ -4,14 +4,24 @@ import 'dart:js_interop_unsafe';
+  external int get arrayLength;
+
+  @JS('at')
+  external JSFunction get elementAt;
```
Overriding `operator[]` didn't help, because then I still get a warning.

pmathew92 replied:
Can this be moved to the js_interop file ?

navaronbracke commented on auth0_flutter/test/web/js_interop_utils_test.dart:
```diff
@@ -27,4 +30,22 @@ void main() {
   });
+
+  group('stripNulls', () {
+    test('removes null values from an object', () {
```
Regression test for the change in stripNulls.

pmathew92 commented on auth0_flutter/lib/auth0_flutter_web.dart:
```diff
@@ -101,22 +107,33 @@ class Auth0Web {
+  ///
+  /// See https://api.dart.dev/dart-js_interop/NullableObjectUtilExtension/jsify.html
+  Future<void> loginWithRedirect({
+    final Object? appState,
```
I think it will be easier if appState is passed as the last option with a default value so it doesn't break for any existing users

navaronbracke replied:
These are _named_ parameters, so the position of the parameter does not matter (you need to specify it as `loginWithRedirect(appState: myStateObject)`)

Since this is already nullable, this does not need a default value (the default _is_ null)

pmathew92 commented on auth0_flutter/lib/auth0_flutter_web.dart:
```diff
@@ -101,22 +107,33 @@ class Auth0Web {
+  }) =>
+      Auth0FlutterWebPlatform.instance.loginWithRedirect(
+        LoginOptions(
+          appState: appState,
```
Same as above . Can we move this as an optional parameter with a default value

navaronbracke replied:
This is an implementation detail. If the `appState` is not passed to the `LoginOptions`, then the feature itself won't work, as the appState is not passed to the Auth0 web SDK.

navaronbracke replied:
LoginOptions.appState is already a named nullable argument, so the position is irrelevant and it has a default (null).

pmathew92 commented on auth0_flutter/lib/src/web/auth0_flutter_plugin_real.dart:
```diff
@@ -37,7 +52,12 @@ class Auth0FlutterPlugin extends Auth0FlutterWebPlatform {
+
+        _appState = JsInteropUtils.dartifyObject(result.appState);
+
+        return;
```
Shouldn't the result be returned here ?

pmathew92 commented on auth0_flutter/lib/src/web/js_interop_utils.dart:
```diff
@@ -4,14 +4,24 @@ import 'dart:js_interop_unsafe';
-      final key = objKeys[i];
+    for (var i = 0; i < objKeys.arrayLength; i++) {
+      // TODO: replace w/ `final key = objKeys[i];` when updating to Dart 3.6.0
+      final key = objKeys.elementAt.callAsFunction(objKeys, i.toJS)!;
```
why would we need this change ? What exactly is this trying to achieve here?

navaronbracke replied:
The `[]` operator and the `length` getter on JSArray are only available from Dart 3.6.0 onwards.
However, requiring to go from Dart 3.5.0 to Dart 3.6.0 would imply a switch from Flutter 3.24 to 3.27.
Therefor I opted to use a backwards compatible fix for now.

pmathew92 commented on auth0_flutter/lib/src/web/js_interop_utils.dart:
```diff
@@ -24,12 +34,30 @@ class JsInteropUtils {
+  /// that do not fit into a static interop definition.
+  ///
+  /// See https://api.dart.dev/dart-js_interop/NullableObjectUtilExtension/jsify.html
+  static JSAny? jsifyObject(final Object? obj) => obj.jsify();
```
@navaronbracke  Do we need this helper methods ? Wouldn't be much easier if we call .jsify() and .dartify() directly on the corresponding objects

navaronbracke commented on auth0_flutter/lib/src/web/js_interop.dart:
```diff
@@ -43,11 +43,25 @@ extension type AuthorizationParams._(JSObject _) implements JSObject {
-  external factory RedirectLoginOptions(
-      {final AuthorizationParams authorizationParams, final String fragment});
+  external factory RedirectLoginOptions({
+    final JSAny? appState,
```
The `RedirectLoginOptions` is a JS interop type and is an implementation detail that is not directly used by users.

navaronbracke commented:
@pmathew92 PTAL. I updated the JS interop to call dartify/jsify directly and moved the polyfill for JSArray to the correct file.

As for the `appState` argument, I think that that is working as intended? Named arguments can be written anywhere in the argument list of a function call. (as opposed to positional arguments, which do have a fixed order). And since the `appState` is nullable users can just omit the argument if it is not needed.

For the switch to Dart 3.6.0, that is up to you, as this would mean that users that are on Flutter 3.24 (for WASM support) will need to migrate to Flutter 3.27 for this change.

pmathew92 commented:
Thankyou @navaronbracke

navaronbracke commented:
Any ideas why the code coverage check is failing? The indirect diff seems to be unrelated to this PR?

Widcket commented:
Thanks @navaronbracke! Appreciate the patience, and the thorough PR.

navaronbracke commented:
Thanks for getting this landed!
